src/System.Management.Automation/engine/NativeCommandParameterBinder.cs

```diff
@@ -418,6 +418,27 @@ private void PossiblyGlobArg(string arg, CommandParameterInternal parameter, boo
             }
 #endif // UNIX
 
+#if !UNIX
+            if (ExperimentalFeature.IsEnabled(ExperimentalFeature.PSNativeWindowsTildeExpansion) && !usedQuotes)
```

Comment:
This is where this feature gets a bit complicated and the original implementation was abandoned.  Double quotes are expected to be expanded (like if your path has whitespace), but single quotes are treated as literal and never expanded.  So this needs to be differentiated.

Reply:
Would you expect tilde expansion if it was surrounded with double quotes? Currently `pwsh` on linux doesn't expand tilde for double quotes:
https://github.com/PowerShell/PowerShell/blob/ac84c35820eeaf46db2d37f8b91ed4b5683c8050/test/powershell/Language/Scripting/NativeExecution/NativeUnixGlobbing.Tests.ps1#L130-L133

Also I checked `bash`, `fish`, `zsh` with `/bin/echo "~/"` and none of them expand it (whereas `/bin/echo ~` is expanded). So I think it's reasonable to just follow every other shell here, are there other considerations for how tilde should behave on windows?

Also in a related area, the tab completion around tilde doesn't seem quite right - the behaviour from powershell 5 seems more correct to me.
Some cases I tried out, where `$` is cursor position:

| Case            | `bash --norc` (5.2)            | `pwsh -NoProfile` (7.4.0) | `powershell -NoProfile` (5.1)    |
| --------------- | ------------------------------ | ------------------------- | -------------------------------- |
| `ls "~/my<tab>` | `ls "/Users/user/my folder"/$` | `ls "~\my folder\$"`      | `ls "C:\Users\user\my folder\$"` |
| `ls ~/my<tab>`  | `ls ~/my\ folder/$`            | `ls '~\my folder\$'`      | `ls 'C:\Users\user\my folder\$'` |

Seems like it is new (undesirable) behaviour in 7.4.0, some discussion about it here: https://github.com/PowerShell/PowerShell/issues/20750

Reply:
POSIX-like shells only ever expand an _unquoted_ argument-initial `~`, either in isolation or if followed by a an _unquoted_ _username_ (_not_ implemented in PowerShell - see #12387) or an _unquoted_ `/`, and then possibly followed by an mix of unquoted and quoted strings (any mix of `'...'`, `"..."`, with their usual semantics), so I think that's what PowerShell should do as well - with the addition of also recognizing unquoted `\`, as stated in the initial post.

Thus, to ensure expansion of `~`-prefixed paths parts of which require quoting, _only the part after `~/`_ must be quoted; e.g. `/bin/echo ~/'a test'` (ditto in PowerShell) or `/bin/echo ~/a\ test` (``/bin/echo ~/a` test`` in PowerShell)

Note that these cases are _currently broken_ in PowerShell (which currently surface on Unix only):
  * #20754
  * They're also broken with respect to _tab-completion_, which inappropriately turns `~/'a test'` into `'~/a test'`, for instance: #20851
 
Indeed, the problem described in #20750  effectively broke tab-completion of `~` on Windows for native-program calls, due to - by well-meaning design - no longer instantly expanding `~` to the value of `$HOME`.

This PR would make that problem go away - but to resolve #20750 this way (which I personally prefer) requires it to become a _stable_ feature right away.

Find a summary of the solution options at:

* https://github.com/PowerShell/PowerShell/issues/20750#issuecomment-1843232925

Reply:
@SteveL-MSFT i think this change just makes the UNIX and Windows behaviors similar, where `~` or `~/` would now behave consistently regardless of platform (it's cribbed from the code in lines 399-418). I would probably suggest `#else` rather than `#endif` at line 419 which _might_ be clearer. We disambiguate between quoted and non-quoted `~` in lines after 344, so that would remain consistent. We don't currently treat `~` with regard to single and double quotes differently; if it's quoted, it's not expanded. 

Now that I think on it, the concern I _do_ have is that we would now have duplicated code which means a fix in one spot may not happen in the other. Perhaps this method should be refactored where we don't duplicate the code.